som-interpreter-bc/src/interpreter.rs

```diff
@@ -195,12 +196,76 @@ impl Interpreter {
                     };
                     let signature = universe.lookup_symbol(symbol);
                     let nb_params = nb_params(signature);
-                    let method = self
-                        .stack
-                        .iter()
-                        .nth_back(nb_params)
-                        .unwrap()
-                        .lookup_method(universe, symbol);
+                    let method = {
+                        let receiver = self.stack.iter().nth_back(nb_params)?;
+                        let receiver_class = receiver.class(universe);
+                        match frame.borrow().kind() {
+                            FrameKind::Block { block } => {
+                                let mut inline_cache_receiver =
+                                    block.inline_cache_receiver.borrow_mut();
+                                let mut inline_cache_invocable =
+                                    block.inline_cache_invocable.borrow_mut();
+                                // SAFETY: this access is actually safe because the bytecode compiler
+                                // makes sure the cache has as many entries as there are bytecode instructions,
+                                // therefore we can avoid doing any redundant bounds checks here.
+                                let maybe_found_receiver = unsafe {
+                                    inline_cache_receiver.get_unchecked_mut(bytecode_idx)
+                                };
+                                let maybe_found_invocable = unsafe {
+                                    inline_cache_invocable.get_unchecked_mut(bytecode_idx)
```

Comment:
The max inline cache size is 1 right? If you used PySOM as a ref, that's what it does indeed, but you might want to investigate a max cache size that's above that. Or not, maybe 1 is already good enough (most call sites will only call a single unique method, but ones that make heavy use of polymorphism won't be happy with a cache that small)
If it's fast as is, then it's fast as is, though! PySOM is doing OK with just 1.

Reply:
PySOM should have 2 possible cache entries `cached_layout1` and `cached_layout2`:
https://github.com/SOM-st/PySOM/blob/d85ed9d957c2210bee10a836dac4454432e6c965/src/som/interpreter/bc/interpreter.py#L704

Reply:
Right, it's 2, my bad... I forgot you used the free BC after send like this

Reply:
I didn't quite notice this during my exploration of PySOM, thanks for pointing this out.  
I'll try to implement something like this, and report on this PR how it changes the performance.